### PR TITLE
Add build tag nomsgpack

### PR DESCRIPTION
**Pull request by sapk**
This PR add a build tag `nomsgpack` to exclude heavy deps `github.com/ugorji/go/codec` that may be not needed for most project. The goal is to reduce the resulting binary size. This could maybe improve some performances but that not the goal so I haven't bench it.

I go for the no impact by default by using a negate tag so that without change it will not impact project depending on this lib.

~~I will try to provide some results of binary size based on https://github.com/gin-gonic/examples.~~ -> see [comment](https://github.com/gin-gonic/gin/pull/1852#issuecomment-481043187)

Should fix (or at least a mitigation to) #1847

### CODE REVIEW

**Comment by sapk** [comment]
Using the basic example : 
```
go build ./
go build -ldflags="-s -w" -o basic.stripped ./
go build -tags nomsgpack -o basic.nomsgpack ./
go build -tags nomsgpack -ldflags="-s -w" -o basic.nomsgpack.stripped ./
ls -l basic*
-rwxr-xr-x 1 sapk sapk 17129377  9 avril 01:17 basic
-rwxr-xr-x 1 sapk sapk 12749851  9 avril 01:18 basic.nomsgpack
-rwxr-xr-x 1 sapk sapk  9342944  9 avril 01:21 basic.nomsgpack.stripped
-rwxr-xr-x 1 sapk sapk 12755840  9 avril 01:22 basic.stripped
```

Without the `nomsgpack` flag the resulting binary file are 35% bigger.

**Comment by zzjin** [comment]
sounds good for ci. any update here?

**Comment by appleboy** [comment]
@sapk Could you resolve the conflicts?

**Comment by sapk** [comment]
@lunny done

**Comment by lunny** [comment]
@sapk :)

**Comment by appleboy** [comment]
@thinkerou what do you think about this PR?

**Comment by sapk** [comment]
To re-validate, I re-do the size bench after last rebase. 
```
-rwxr-xr-x 1 sapk sapk 15269181 28 déc.  19:41 basic
-rwxr-xr-x 1 sapk sapk 12925564 28 déc.  19:40 basic.nomsgpack
-rwxr-xr-x 1 sapk sapk  9506816 28 déc.  19:41 basic.nomsgpack.stripped
-rwxr-xr-x 1 sapk sapk 11239424 28 déc.  19:40 basic.stripped

```

File: binding/binding_msgpack_test.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
+// Copyright 2014 Manu Martinez-Almeida.  All rights reserved.
```
Comment (thinkerou) [inline binding/binding_msgpack_test.go]: please copyright

File: binding/binding_nomsgpack.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
+// Copyright 2014 Manu Martinez-Almeida.  All rights reserved.
```
Comment (thinkerou) [inline binding/binding_nomsgpack.go]: 1. please update copyright
2. filename: `binding_nomsgpack.go` `binding_msgpack_test.go`, should we use `xxx_msgpack_xxx`?
3. the file and `binding.go` have repeat code, should we use reduce?
thanks!

File: binding/binding_msgpack_test.go (hunk 1 above)
Comment (sapk) [reply binding/binding_msgpack_test.go]:
```suggestion
// Copyright 2020 Gin Core Team. All rights reserved.
```

File: binding/binding_nomsgpack.go (hunk 2 above)
Comment (sapk) [reply binding/binding_nomsgpack.go]:
```suggestion
// Copyright 2020 Gin Core Team. All rights reserved.
```

I will have fresh look on removing/refactor duplicate code later.

**Comment by appleboy** [comment]
maybe need another PR to update documentation.

**Comment by sapk** [comment]
@lunny for the doc you mean in the README.md ?

**Comment by appleboy** [comment]
@sapk like as https://github.com/gin-gonic/gin#build-with-jsoniter

**Pull request merged**